File: Kattis/NAC 2024/Arrested_Development.py
import sys
from collections import defaultdict
input = sys.stdin.readline
def multiple_ints():
  return map(int, input().strip().split())

def solve():  
  n = int(input())

  times = []

  for i in range(n):
    a,b = multiple_ints()
    times.append((a,b))

  MX = 10**10
  
  dp = defaultdict(lambda: MX)

  #let dp[i][j] = best time we can get on person 2 assuming person 1 has used J time up until I

  dp[0] = times[0][1]
  dp[times[0][0]] = 0

  for i in range(1, n):
    ndp = defaultdict(lambda: MX)
    a,b = times[i]
    for prior_A_time, best_b_conditioned_on_A in dp.items():
      #take nums[i] for A or don't
      
      #take
      ndp[prior_A_time+a] = min(ndp[prior_A_time+a], best_b_conditioned_on_A)

      #give to B
      ndp[prior_A_time] = min(ndp[prior_A_time], best_b_conditioned_on_A + b)
    dp = ndp
  

  
  ans = MX

  for a_time, best_b_conditioned_on_A in dp.items():
    ans = min(ans, max(a_time, best_b_conditioned_on_A))
  print(ans)


solve()
    

File: Kattis/NAC 2024/Arrested_Development.cpp
#include <bits/stdc++.h>
using namespace std;
void solve(){
  int n;
  cin >> n;
  vector<pair<int, int>> times;
  int MX = pow(10, 8);
  int aT = 0;
  for(int i = 0; i < n; i++){
    int a; int b; cin >> a >> b;
    aT += a;
    times.emplace_back(a,b);
  }

  vector<vector<int>> dp(n, vector<int>(aT+1, MX));

  dp[0][times[0].first] = 0;
  dp[0][0] = times[0].second;

  for(int i = 1; i < n; i++){
    for(int j = 0; j <= aT; j++){
      //take it for intern B
      dp[i][j] = dp[i-1][j] + times[i].second;
      //take it for intern A
      if(j >= times[i].first){
        dp[i][j] = min(dp[i][j], dp[i-1][j-times[i].first]);
      }
    }
  }
  int ans = MX;
  for(int i = 0; i <= aT; i++){
    //cout << i << " H E R E " << dp[n-1][i] << endl;
    ans = min(ans, max(i, dp[n-1][i]));
  }
  cout << ans;
}

int main(){
  solve();
  return 0;
}